File: drivers/java/src/main/java/dev/reddb/helpers/TxClient.java
package dev.reddb.helpers;

/**
 * Implements {@code tx.*} from the SDK Helper Spec. Supports both
 * imperative ({@link #begin}/{@link #commit}/{@link #rollback}) and
 * callback ({@link #run}) forms. Nested {@code run} calls reject with
 * {@code INVALID_ARGUMENT} — callers needing savepoints issue
 * {@code SAVEPOINT}/{@code RELEASE} via the underlying connection.
 */
public final class TxClient {
    private final Querier q;
    private boolean inCallback;

    TxClient(Querier q) { this.q = q; }

    /** Open a transaction. */
    public void begin() {
        q.query("BEGIN");
    }

    /** Commit the current transaction. */
    public void commit() {
        q.query("COMMIT");
    }

    /** Roll back the current transaction. */
    public void rollback() {
        q.query("ROLLBACK");
    }

    @FunctionalInterface
    public interface TxBody {
        void apply(TxClient tx) throws Exception;
    }

    /**
     * Run {@code body} inside a transaction. The body's return path
     * commits; any thrown exception rolls back and is re-thrown wrapped
     * in {@link RuntimeException} if checked.
     */
    public void run(TxBody body) {
        if (inCallback) {
            throw new HelperException.InvalidArgument(
                "tx.run does not support nesting; use SAVEPOINT via conn.query directly");
        }
        inCallback = true;
        try {
            begin();
            try {
                body.apply(this);
            } catch (RuntimeException re) {
                safeRollback();
                throw re;
            } catch (Exception checked) {
                safeRollback();
                throw new RuntimeException(checked);
            }
            commit();
        } finally {
            inCallback = false;
        }
    }

    private void safeRollback() {
        try { rollback(); } catch (RuntimeException ignored) { /* best effort */ }
    }
}

File: drivers/php/src/Helpers/TxClient.php
<?php
/**
 * Implements `tx.*` from the SDK Helper Spec (v1.0 §7) — imperative form
 * (`begin` / `commit` / `rollback`) plus the optional callback form
 * (`run`). Mirrors `drivers/go/helpers.go` (TxClient).
 *
 * The connection is session-stateful: a `begin` opens a transaction that the
 * next `commit` or `rollback` closes. Concurrent calls on the same client
 * during an open transaction MUST serialise (the underlying transport
 * already serialises one in-flight statement at a time).
 */

declare(strict_types=1);

namespace Reddb\Helpers;

final class TxClient
{
    public function __construct(
        private readonly Querier $q,
        private readonly bool $inTx = false,
    ) {}

    /** Start a transaction. Returns the raw JSON envelope. */
    public function begin(): string
    {
        return $this->q->query('BEGIN');
    }

    /** Commit the open transaction. Returns the raw JSON envelope. */
    public function commit(): string
    {
        return $this->q->query('COMMIT');
    }

    /** Roll back the open transaction. Returns the raw JSON envelope. */
    public function rollback(): string
    {
        return $this->q->query('ROLLBACK');
    }

    /**
     * Run a SQL statement inside the transaction session. Provided so a
     * `run` callback can issue savepoints or arbitrary statements without
     * reaching back to the connection.
     *
     * @param array<int,mixed> $params
     */
    public function query(string $sql, array $params = []): string
    {
        return $this->q->query($sql, $params);
    }

    /**
     * Callback form (spec §7.2): begin, invoke `$fn`, then commit on success
     * or rollback + re-throw on failure. Nested `run` is rejected with
     * INVALID_ARGUMENT — the PHP driver does NOT use savepoints; callers
     * wanting nested semantics issue `SAVEPOINT` directly via {@see query()}.
     *
     * @param callable(TxClient):void $fn
     */
    public function run(callable $fn): void
    {
        if ($this->inTx) {
            throw new InvalidArgument(
                'tx.run does not support nested transactions; use SAVEPOINT explicitly'
            );
        }
        $this->begin();
        $child = new self($this->q, true);
        try {
            $fn($child);
        } catch (\Throwable $e) {
            try {
                $this->rollback();
            } catch (\Throwable) {
                // Surface the original callback failure, not the rollback noise.
            }
            throw $e;
        }
        $this->commit();
    }
}
